### PR TITLE
Updating Rails Engine Project Requirements

### PR DESCRIPTION
**Pull request by iandouglas**
Removed most CRUD endpoints for merchants, added several new non-RESTful routes, added a note about returning error structures, and noting that students will get to choose which endpoints they want to build by making some choices.

I'll update the text later to use a Postman endpoint collection including tests instead of Rails Driver. At a minimum, I'll update Rails Driver (and any necessary database data in the pg_dump file), but I'd sure like to use a Postman collection going forward.

### CODE REVIEW

File: module3/projects/rails_engine/requirements.md [hunk 1]
```diff
@@ -93,17 +99,86 @@ module Api
 end
 ```
 
-## 3a. ReST Endpoints
+#### Error Responses
+
+If the user causes an error for which you are sending a 400-series error code, the JSON body of the response should follow a similar JSON Spec of a `data` element, with a `nil` ID, and empty attributes.
+
+As an **EXTENSION**, customize the error message to use this format instead:
```
Comment (dionew1) [inline module3/projects/rails_engine/requirements.md:103]: Do we want this extension to also follow the JSON Spec with a `data` and have message and errors be nested in attributes?

File: module3/projects/rails_engine/requirements.md [hunk 2]
```diff
@@ -93,17 +99,86 @@ module Api
...
+  * find a quantity of merchants sorted by descending items sold
+  * total revenue generated in the whole system over a start/end date range
+  * total revenue for a given merchant
+  * find a quantity of items sorted by descending revenue
+  * total revenue of successful invoices which have not yet been shipped
+  * revenue report, broken down by month in ascending date order
+
+In total, the MINIMUM requirement will be 15 endpoints.
+
+**HINT**: Invoices must have a successful transaction and shipped to the customer to be considered as revenue.
```
Comment (dionew1) [inline module3/projects/rails_engine/requirements.md:157]: Also as part of a hint do we want to tell them for the start/end date range they should use the dates on the invoice table?

File: module3/projects/rails_engine/requirements.md [hunk 3]
```diff
@@ -135,115 +210,177 @@ Example JSON response for the Merchant resource:
...
-      "name": "Store Name"
+      "name": "Widget",
+      "description": "High quality widget",
+      "unit_price": 100.99,
+      "created_at": "2020-12-18T13:42:56 UTC",
+      "updated_at": "2020-12-18T13:42:56 UTC"
     }
   }
 }
 ```
```
Comment (dionew1) [inline module3/projects/rails_engine/requirements.md:274]: Could we ask them to return the same data as fetching a single item, leaving off the created_at and updated_at so they can reuse the same serializer? Also, should we ask them to specify the status code of 201 for the response?

File: module3/projects/rails_engine/requirements.md [hunk 4]
```diff
@@ -135,115 +210,177 @@ Example JSON response for the Merchant resource:
...
     "id": "1",
     "type": "merchant",
     "attributes": {
-      "name": "Store Name"
+      "name": "New Widget Name",
+      "description": "High quality widget, now with more widgety-ness",
+      "unit_price": 299.99,
+      "created_at": "2020-12-18T13:42:56 UTC",
+      "updated_at": "2020-12-20T18:12:45 UTC"
     }
```
Comment (dionew1) [inline module3/projects/rails_engine/requirements.md:308]: Similar to the above create comment, or do we want them to see the time differences?

File: module3/projects/rails_engine/requirements.md [hunk 5]
```diff
@@ -135,115 +210,177 @@ Example JSON response for the Merchant resource:
...
-
-Example JSON response for `GET /api/v1/merchants/find?name=ring`
+  * return a 404 if merchant is not found
+* `GET /api/v1/items/:id/merchant` - return the merchant associated with an item
+  * return a 404 if the item is not found
+  
+
+## Non-RESTful Search Endpoints
+
+In addition to the standard RESTful endpoints described above, you will build the following endpoints which will NOT follow RESTful convention:
```
Comment (dionew1) [inline module3/projects/rails_engine/requirements.md:334]: To be consistent with messaging, maybe update to `you will build 1 pair of the following endpoints'

**Review by dionew1 — COMMENTED** [review]
I like the changes and giving students some choice about what endpoints to complete! I also think the Postman test suite will be great to further emphasize to students how this tool can be used and hopefully get them to be comfortable with it. My only concern is that 15 points are going to be a lot for them to try to complete on top of expectations for error handling in a 1 week project. 

I think there are a few solutions for my concern that I'm going to throw out to get your thoughts. 

1. We keep it as 15 endpoints, but make the timeline 1.5 weeks. We adjust the pair project to also be 1.5 weeks. I think having them in groups of 3 this inning allowed them to work through it quicker and still achieve the same amount of learning, so maybe rather than pair we make it 3 ppl?

2. We adjust the number of required endpoints to be 11 endpoints and have the additional be part of extensions. I could see the find endpoints and relationship endpoints being moved to the extension category to encourage them to focus on the RESTful and the AR.

3. We keep it as it is, give it a go and see if this concern is something that even needs to be addressed. 😄

File: module3/projects/rails_engine/requirements.md (hunk 1 above)
Comment (iandouglas) [reply module3/projects/rails_engine/requirements.md:103]: No, typical APIs would have a completely different structure for error messages

File: module3/projects/rails_engine/requirements.md (hunk 2 above)
Comment (iandouglas) [reply module3/projects/rails_engine/requirements.md:157]: no, I think that's something they can still discover on their own

File: module3/projects/rails_engine/requirements.md (hunk 3 above)
Comment (iandouglas) [reply module3/projects/rails_engine/requirements.md:274]: ah yes, I meant to remove the timestamps from these

File: module3/projects/rails_engine/requirements.md (hunk 5 above)
Comment (iandouglas) [reply module3/projects/rails_engine/requirements.md:334]: Yes, good catch

**Comment by iandouglas** [comment]
Thanks @dionew1 

I think the idea of the error messaging was maybe too brief, as I do say to send back an empty response with a 400-series error and as an extension to make a custom error response. Maybe moving the custom error information to the extensions area would be best here.

Right now they're building 20 endpoints:
- 5 CRUD for Merchants
- 5 CRUD for Items
- 2 relationship endpoints (items for a merchant, merchant for an item)
- 2 searches for Merchants
- 2 searches for Items
- 4 business intelligence

I lowered it to 15, I think maybe I just spell things out a little more clearly for them...
- 2 for Merchants, show and index, and index has pagination, 3 fewer than before
- 5 CRUD for Items, and index has pagination (same as before)
- 2 relationship endpoints (same as before)
- 1 search for Merchants (one fewer)
- 1 search for Items (one fewer)
- 4 business intelligence (same as before)

If anything, we could bump up the business intelligence to 5 or 6 endpoints. I'd also be okay removing the update endpoint for Items too.

**Pull request merged**